pkg/streamingpromql/operators/aggregations/topkbottomk/range_query.go

```diff
@@ -0,0 +1,505 @@
+	data := types.InstantVectorSeriesData{}
+
+	if series.pointCount == 0 {
+		return data, nil
+	}
+
+	var err error
+	data.Floats, err = types.FPointSlicePool.Get(series.pointCount, t.MemoryConsumptionTracker)
+	if err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
+
+	for idx, shouldReturn := range series.shouldReturnPoint {
+		if !shouldReturn {
+			continue
+		}
+
+		data.Floats = append(data.Floats, promql.FPoint{T: t.TimeRange.IndexTime(int64(idx)), F: series.values[idx]})
+	}
+
+	return data, nil
+}
+
+func (t *RangeQuery) readNextSeries(ctx context.Context) error {
+	nextSeries, err := t.Inner.NextSeries(ctx)
+	if err != nil {
+		return err
+	}
+
+	if len(nextSeries.Histograms) > 0 {
+		t.emitIgnoredHistogramsAnnotation()
+	}
+
+	// topk() and bottomk() ignore histograms, so return the HPoint slice to the pool now.
+	types.HPointSlicePool.Put(nextSeries.Histograms, t.MemoryConsumptionTracker)
+
+	g := t.remainingInnerSeriesToGroup[0]
+	t.remainingInnerSeriesToGroup = t.remainingInnerSeriesToGroup[1:]
+	if err := t.accumulateIntoGroup(nextSeries, g); err != nil {
+		return err
+	}
+
+	types.FPointSlicePool.Put(nextSeries.Floats, t.MemoryConsumptionTracker)
+
+	return nil
+}
+
+func (t *RangeQuery) emitIgnoredHistogramsAnnotation() {
+	if t.haveEmittedIgnoredHistogramsAnnotation {
+		return
+	}
+
+	if t.IsTopK {
+		t.annotations.Add(annotations.NewHistogramIgnoredInAggregationInfo("topk", t.expressionPosition))
+	} else {
+		t.annotations.Add(annotations.NewHistogramIgnoredInAggregationInfo("bottomk", t.expressionPosition))
+	}
+
+	t.haveEmittedIgnoredHistogramsAnnotation = true
+}
+
+func (t *RangeQuery) accumulateIntoGroup(data types.InstantVectorSeriesData, g *rangeQueryGroup) error {
+	groupSeriesIndex := g.seriesRead()
+
+	if g.seriesForTimestamps == nil {
+		var err error
+		g.seriesForTimestamps, err = intSliceSlicePool.Get(t.TimeRange.StepCount, t.MemoryConsumptionTracker)
+		if err != nil {
+			return err
+		}
+
+		g.seriesForTimestamps = g.seriesForTimestamps[:t.TimeRange.StepCount]
+	}
+
+	if g.series == nil {
+		var err error
+		g.series, err = rangeQuerySeriesSlicePool.Get(g.totalSeries, t.MemoryConsumptionTracker)
+		if err != nil {
+			return err
+		}
+	}
+
+	g.series = g.series[:groupSeriesIndex+1]
+	thisSeries := &g.series[groupSeriesIndex]
+
+	for _, p := range data.Floats {
+		timestampIndex := t.TimeRange.PointIndex(p.T)
+		limit := t.k[timestampIndex]
+
+		if limit == 0 {
+			continue
+		}
+
+		if g.seriesForTimestamps[timestampIndex] == nil {
+			// This is the first time we've seen a point for this timestamp, create the list of source series.
+			maximumPossibleSeries := min(limit, int64(g.totalSeries))
+
+			var err error
+			g.seriesForTimestamps[timestampIndex], err = types.IntSlicePool.Get(int(maximumPossibleSeries), t.MemoryConsumptionTracker)
+			if err != nil {
+				return err
+			}
+		}
+
+		populateSeriesSlices := func() error {
```

Comment:
Sorry if I'm missing something obvious, but does this need to be inside the for loop?

Reply:
Good catch, yep, this can move outside the loop.